### PR TITLE
Fix/issue 726 emr autoscaling

### LINKED ISSUE #726: EMR: AutoScalingPolicy property missing from InstanceGroupConfigProperty
InstanceGroupConfigProperty is missing AutoScalingPolicy

### PR DESCRIPTION
**Pull request by bmurtagh**
Introduces AutoScaling for EMR clusters found under http://boto3.readthedocs.io/en/latest/reference/services/emr.html

Once merged, this should close #726

### CODE REVIEW

**Comment by markpeek** [comment]
While I believe you're still developing this change (and thank you!), it would be best to revert the last change which renames CloudWatchAlarmDefinition. Where ever possible I try to use the same names from the CF docs for compatibility. A shortened name might cause issues in the future when I finally get the code to auto-generate these classes from the CF definitions.

**Comment by bmurtagh** [comment]
@markpeek I'm in the same school of thought as you and started off that way, but the pycodestyle is failing the tests due to character limits. The indentation plus the longer CloudFormation property names is making this extremely difficult and turning it into a hassle.

**Comment by markpeek** [comment]
I'd say revert to the longer name and let me know. I can then pull it and take a look at the pycodestyle issues.

**Comment by bmurtagh** [comment]
@phobologic @markpeek, any suggestions on how I can get the Python 3.x builds passing? The build in TravisCI for Python 2.7 passes, but all 3.x failed.

**Comment by bmurtagh** [comment]
@markpeek & @phobologic -- thank you for  for the assistance with the Python 3 testing!! We finally got there :)

Looking forward to getting this merged.

File: troposphere/emr.py [hunk 1]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+    }
+
+
 class InstanceGroupConfigProperty(AWSProperty):
     props = {
         'BidPrice': (basestring, False),
         'Configurations': ([Configuration], False),
         'EbsConfiguration': (EbsConfiguration, False),
-        'InstanceCount': (positive_integer, True),
+        'InstanceCount': (positive_integer, False),
```
Comment (phobologic) [inline troposphere/emr.py]: From the docs it looks like this is required: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-emr-cluster-jobflowinstancesconfig-instancegroupconfig.html

File: troposphere/emr.py [hunk 2]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
     }
 
 
+class Constraints(AWSProperty):
```
Comment (phobologic) [inline troposphere/emr.py]: This should be named `ScalingConstraints`

File: troposphere/emr.py [hunk 3]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+    props = {
+        'MinCapacity': (integer, False),
```
Comment (phobologic) [inline troposphere/emr.py]: These are both required per the docs: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-emr-cluster-jobflowinstancesconfig-instancegroupconfig-autoscalingpolicy-constraints-scalingconstraints.html

File: troposphere/emr.py [hunk 4]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+class RulesActionConfig(AWSProperty):
+    props = {
+        'Market': (market_validator, False),
+        'SimpleScalingPolicyConfiguration': (
+            SimpleScalingPolicyConfiguration, False
+        )
+    }
+
+
+class Rules(AWSProperty):
```
Comment (phobologic) [inline troposphere/emr.py]: This should be called `ScalingRule` per the docs: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-emr-cluster-jobflowinstancesconfig-instancegroupconfig-autoscalingpolicy-constraints-scalingrule.html

File: troposphere/emr.py [hunk 5]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+    props = {
+        'Name': (basestring, False),
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py [hunk 6]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+        'Description': (basestring, False),
+        'Action': (RulesActionConfig, False),
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py [hunk 7]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+        'Trigger': (Trigger, False)
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py [hunk 8]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+
+class SimpleScalingPolicyConfiguration(AWSProperty):
+    props = {
+        'AdjustmentType': (basestring, False),
+        'ScalingAdjustment': (positive_integer, False),
+        'CoolDown': (positive_integer, False)
+    }
+
+
+class RulesActionConfig(AWSProperty):
```
Comment (phobologic) [inline troposphere/emr.py]: It looks like this should be called `ScalingAction`: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticmapreduce-cluster-scalingaction.html

File: troposphere/emr.py [hunk 9]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+    props = {
+        'Market': (market_validator, False),
+        'SimpleScalingPolicyConfiguration': (
+            SimpleScalingPolicyConfiguration, False
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py [hunk 10]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+class Trigger(AWSProperty):
+    props = {
+        'CloudWatchAlarmDefinition': (CloudWatchAlarmDefinition, False),
+    }
+
+
+class SimpleScalingPolicyConfiguration(AWSProperty):
+    props = {
+        'AdjustmentType': (basestring, False),
+        'ScalingAdjustment': (positive_integer, False),
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py [hunk 11]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+class Constraints(AWSProperty):
+    props = {
+        'MinCapacity': (integer, False),
+        'MaxCapacity': (integer, False)
+    }
+
+
+class CloudWatchAlarmDefinition(AWSProperty):
+    props = {
+        'ComparisonOperator': (basestring, False),
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py [hunk 12]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+        'EvaluationPeriods': (positive_integer, False),
+        'MetricName': (basestring, False),
+        'Namespace': (basestring, False),
+        'Period': (positive_integer, False),
+        'Statistic': (basestring, False),
+        'Threshold': (positive_integer, False),
+        'Unit': (basestring, False),
+        'Dimensions': ([KeyValue], False)
```
Comment (phobologic) [inline troposphere/emr.py]: This should be `MetricDimension` though it looks like that's mostly a copy of the KeyValue class: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-emr-cluster-jobflowinstancesconfig-instancegroupconfig-autoscalingpolicy-constraints-scalingrule-scalingtrigger-cloudwatchalarmdefinition-metricdimension.html

File: troposphere/emr.py [hunk 13]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+        'MinCapacity': (integer, False),
+        'MaxCapacity': (integer, False)
+    }
+
+
+class CloudWatchAlarmDefinition(AWSProperty):
+    props = {
+        'ComparisonOperator': (basestring, False),
+        'EvaluationPeriods': (positive_integer, False),
+        'MetricName': (basestring, False),
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py [hunk 14]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+        'Namespace': (basestring, False),
+        'Period': (positive_integer, False),
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py [hunk 15]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+        'Statistic': (basestring, False),
+        'Threshold': (positive_integer, False),
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py [hunk 16]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+        'Unit': (basestring, False),
+        'Dimensions': ([KeyValue], False)
+    }
+
+
+class Trigger(AWSProperty):
```
Comment (phobologic) [inline troposphere/emr.py]: Should be called `ScalingTrigger`: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticmapreduce-cluster-scalingtrigger.html

File: troposphere/emr.py [hunk 17]
```diff
@@ -122,15 +122,76 @@ class EbsConfiguration(AWSProperty):
...
+    props = {
+        'CloudWatchAlarmDefinition': (CloudWatchAlarmDefinition, False),
```
Comment (phobologic) [inline troposphere/emr.py]: This is required.

File: troposphere/emr.py (hunk 12 above)
Comment (bmurtagh) [inline troposphere/emr.py]: On:

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-emr-cluster-jobflowinstancesconfig-instancegroupconfig-autoscalingpolicy-constraints-scalingrule-scalingtrigger-cloudwatchalarmdefinition.html#cfn-emr-cluster-jobflowinstancesconfig-instancegroupconfig-autoscalingpolicy-constraints-scalingrule-scalingtrigger-cloudwatchalarmdefinition-dimensions

I see the the property listed as `Dimensions` which is why I used that term, but under the above URL under the Dimensions section, there's a link to the `MetricDimension`. Which one should I use?

**Comment by bmurtagh** [comment]
@phobologic @markpeek are my updated changes able to be approved?

**Comment by markpeek** [comment]
Thanks!

**Pull request merged**